### PR TITLE
Don't write to config file if `config_is_read_only` is set

### LINKED ISSUE #29901: Config setting `config_is_read_only` should prevent Nextcloud from writing to `config.php`
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Description

The config setting `config_is_read_only` currently only prevents Nextcloud from checking for writable config in server health checks.

That's also how the setting is described at https://github.com/nextcloud/documentation/blob/6efc785734a23786bff4c6340a03fa2acbf12f42/admin_manual/configuration_server/config_sample_php_parameters.rst#nextcloud-verifications.

Still, users/admins expect the setting to prevent Nextcloud from writing to `config.php`. That would be helpful, e.g. to prevent admin users from accidently changing server configuration.

### Steps to reproduce
1. Set `'config_is_read_only' => true` in `config.php`
2. Make changes to Nextcloud config via admin interface or `occ`
3. Nextcloud writes to `config.php`

### PR DESCRIPTION
**Pull request by mejo-**
Also don't write to cache in this case to prevent cache and config file
being out of sync.

Fixes: #29901

### CODE REVIEW

**Comment by mejo-** [comment]
I opened a corresponding PR in nextcloud/documentation: https://github.com/nextcloud/documentation/pull/7795

**Comment by nickvergessen** [comment]
So this means all the paths which set a config need to check for the exception?

**Comment by mejo-** [comment]
> So this means all the paths which set a config need to check for the exception?

Right now, a config file with read-only permissions throws the same kind of exception. So in my eyes, adding the check is not a regression to how it was before. What do you think?

I agree that proper notifications about the config file being set to read-only would be nice in the long term. But that can be tackled in another PR later, no?

**Comment by acsfer** [comment]
Does it make sense to use `OC::$CLI` so writes would always be possible using the CLI?

```
if (OC::$CLI) {
  // write
} else {
  // exception
}
```

**Comment by nickvergessen** [comment]
That also is true for cron jobs where I would not expects write to happen.

File: lib/private/Config.php [hunk 1]
```diff
@@ -109,6 +112,7 @@ public function getValue($key, $default = null) {
 	 *
 	 * @param array $configs Associative array with `key => value` pairs
 	 *                       If value is null, the config key will be deleted
+	 * @throws HintException
```
Comment (ChristophWurst) [inline lib/private/Config.php:115]: shouldn't this be on the interface? or is this just some internal knowledge and we expect the calling site to fail hard when they run into this?

Comment (ChristophWurst) [reply lib/private/Config.php:115]: ah, @nickvergessen already asked about this: https://github.com/nextcloud/server/pull/30130#issuecomment-987830088

Comment (mejo-) [reply lib/private/Config.php:115]: You're right, the interface should have the annotation as well. Added now. Is that, what you meant?

**Comment by mejo-** [comment]
> So this means all the paths which set a config need to check for the exception?

Do you mean that throwing the exception should be exposed on the interface (`IConfig`) as well, similar to @ChristophWurst's comment? If so, then I added this now: https://github.com/nextcloud/server/pull/30130/files#diff-54875ef8945bf07212017eccde83d6400812202ce1e30227e9fda9e92fa9474d

File: lib/private/Config.php (hunk 1 above)
Comment (ChristophWurst) [reply lib/private/Config.php:115]: Yes.

Comment (mejo-) [reply lib/private/Config.php:115]: Perfect, thanks :blush: Anything else you'd like to change before approving the PR?

**Review by julien-nc — APPROVED** [review]

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
Leaving the merge to a re-review from @ChristophWurst or @nickvergessen

**Review by ChristophWurst — APPROVED** [review]
:shipit:

**Comment by ChristophWurst** [comment]
@mejo- please add this to https://github.com/nextcloud/server/issues/29914 *and* update the developer docs about the change :v:

**Comment by mejo-** [comment]
Yay, thanks @ChristophWurst!

Added to #29914. Docs should be updated automatically with this PR. At least @nickvergessen said that `admin_manual/configuration_server/config_sample_php_parameters.rst` will be auto-generated from `config/config.sample.php` in https://github.com/nextcloud/documentation/pull/7795. Or what do you mean with "developer docs"?

**Comment by ChristophWurst** [comment]
The admin docs will be updated with the sample config, yes. But https://docs.nextcloud.com/server/latest/developer_manual/basics/storage/configuration.html should receive an update about the new exception

**Comment by mejo-** [comment]
> The admin docs will be updated with the sample config, yes. But https://docs.nextcloud.com/server/latest/developer_manual/basics/storage/configuration.html should receive an update about the new exception

Ah, thanks for the pointer! Did this now: https://github.com/nextcloud/documentation/pull/7851

**Pull request merged**